File: 質數/歐拉篩法.java
import java.io.*;
import java.util.*;

public class Main {
    static int MAX_NUM = 1000000;
    static int[] mark = new int[(MAX_NUM >> 6) + 1];
	static ArrayList<Integer> primes = new ArrayList<>();
	public static void main(String[] args) throws IOException {
	    eulerSieve();
	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StreamTokenizer st = new StreamTokenizer(br);
        StringBuilder output = new StringBuilder();
        int n;
        while(st.nextToken() == StreamTokenizer.TT_NUMBER && (n = (int)st.nval) != 0)
            output.append(n).append(n == 2 || ((n & 1) == 1 && !GET(n >> 1)) ? " is prime\n" : " is not prime\n");
        System.out.print(output);
	}
	
	static boolean GET(int n) {
	    return (mark[n >> 5] & (1 << (n & 31))) != 0;
	}
	
	static void SET(int n) {
	    mark[n >> 5] |= 1 << (n & 31);
	}
	
	static void eulerSieve() {
	    SET(0);
	    primes.add(2);
        for(int n = 3; n <= MAX_NUM; n += 2) {
            if(!GET(n >> 1))
                primes.add(n);
            for(int i = 1, temp; (temp = primes.get(i) * n) <= MAX_NUM; ++i) {
                SET(temp >> 1);
                if(n % primes.get(i) == 0)
                    break;
            }
        }
    }
}

File: 質數/歐拉篩法.cpp
#include <bits/stdc++.h>
using namespace std;

const int MAX_NUM = 10000000;
bitset<(MAX_NUM >> 1) + 1> mark;
vector<int> primes;

void eulerSieve() {
    mark[0] = 1;
    primes.push_back(2);
    for(int n = 3; n <= MAX_NUM; n += 2) {
        if(!mark[n >> 1])
            primes.push_back(n);
        for(int i = 1, temp; (temp = primes[i] * n) <= MAX_NUM; ++i) {
            mark[temp >> 1];
            if(n % primes[i] == 0)
                break;
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    eulerSieve();
    int n;
    while(cin >> n)
        cout << n << (n == 2 || (n & 1 && !mark[n >> 1]) ? " is prime\n" : " is not prime\n"); 
    return 0;
}
